File: console/src/main/scala/net/kemitix/thorp/console/ConsoleOut.scala
package net.kemitix.thorp.console

import net.kemitix.thorp.domain.StorageEvent.ActionSummary
import net.kemitix.thorp.domain.Terminal._
import net.kemitix.thorp.domain.{Bucket, RemoteKey, Sources}
import zio.UIO

import scala.io.AnsiColor._

sealed trait ConsoleOut {
  def en: String
}

object ConsoleOut {

  sealed trait WithBatchMode {
    def en: String
    def enBatch: String
    def apply(batchMode: Boolean): UIO[String] =
      selectLine(batchMode)
    private def selectLine(batchMode: Boolean) =
      if (batchMode) UIO(enBatch) else UIO(en)
  }

  final case class ValidConfig(
      bucket: Bucket,
      prefix: RemoteKey,
      sources: Sources
  ) extends ConsoleOut {
    private val sourcesList = sources.paths.mkString(", ")
    override def en: String =
      List(s"Bucket: ${bucket.name}",
           s"Prefix: ${prefix.key}",
           s"Source: $sourcesList")
        .mkString(", ")
  }

  final case class UploadComplete(remoteKey: RemoteKey)
      extends ConsoleOut.WithBatchMode {
    override def en: String =
      s"${GREEN}Uploaded:$RESET ${remoteKey.key}$eraseToEndOfScreen"
    override def enBatch: String =
      s"Uploaded: ${remoteKey.key}"
  }

  final case class CopyComplete(sourceKey: RemoteKey, targetKey: RemoteKey)
      extends ConsoleOut.WithBatchMode {
    override def en: String =
      s"${GREEN}Copied:$RESET ${sourceKey.key} => ${targetKey.key}$eraseToEndOfScreen"
    override def enBatch: String =
      s"Copied: ${sourceKey.key} => ${targetKey.key}"
  }

  final case class DeleteComplete(remoteKey: RemoteKey)
      extends ConsoleOut.WithBatchMode {
    override def en: String =
      s"${GREEN}Deleted:$RESET ${remoteKey.key}$eraseToEndOfScreen"
    override def enBatch: String =
      s"Deleted: $remoteKey"
  }

  final case class ErrorQueueEventOccurred(action: ActionSummary, e: Throwable)
      extends ConsoleOut.WithBatchMode {
    override def en: String =
      s"${RED}ERROR:$RESET ${action.name} ${action.keys}: ${e.getMessage}$eraseToEndOfScreen"
    override def enBatch: String =
      s"${action.name} failed: ${action.keys}: ${e.getMessage}"
  }
}

File: console/src/main/java/net/kemitix/thorp/console/ConsoleOut.java
package net.kemitix.thorp.console;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import net.kemitix.thorp.domain.*;
import scala.io.AnsiColor;

import java.nio.file.Path;
import java.util.Arrays;
import java.util.function.Function;
import java.util.stream.Collectors;

public interface ConsoleOut {
    String en();
    default String eraseToEndOfScreen() {
        return Terminal.eraseToEndOfScreen;
    }
    default String reset() {
        return "\u001B[0m";
    }
    default String red() {
        return "\u001B[31m";
    }
    default String green() {
        return "\u001B[32m";
    }
    interface WithBatchMode extends ConsoleOut, Function<Boolean, String> {
        String enBatch();
        default String selectLine(boolean batchMode) {
            return batchMode
                    ? enBatch()
                    : en();
        }
        @Override
        default String apply(Boolean batchMode) {
            return selectLine(batchMode);
        }
    }

    static ConsoleOut validConfig(
            Bucket bucket,
            RemoteKey prefix,
            Sources sources
    ) {
        return new ValidConfig(bucket, prefix, sources);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class ValidConfig implements ConsoleOut {
        private final Bucket bucket;
        private final RemoteKey prefix;
        private final Sources sources;

        @Override
        public String en() {
            return String.join(", ", Arrays.asList(
                    "Bucket: " + bucket.name(),
                    "Prefix: " + prefix.key(),
                    "Source: " + sources.paths().stream()
                            .map(Path::toString)
                            .collect(Collectors.joining(", "))));
        }
    }

    static ConsoleOut.WithBatchMode uploadComplete(RemoteKey remoteKey) {
        return new UploadComplete(remoteKey);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class UploadComplete implements ConsoleOut.WithBatchMode {
        private final RemoteKey remoteKey;
        @Override
        public String en() {
            return String.format("%sUploaded:%s %s%s",
                    green(), reset(),
                    remoteKey.key(),
                    eraseToEndOfScreen());
        }
        @Override
        public String enBatch() {
            return String.format("Uploaded: %s", remoteKey.key());
        }
    }

    static ConsoleOut.WithBatchMode copyComplete(RemoteKey sourceKey, RemoteKey targetKey) {
        return new CopyComplete(sourceKey, targetKey);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class CopyComplete implements ConsoleOut.WithBatchMode {
        private final RemoteKey sourceKey;
        private final RemoteKey targetKey;
        @Override
        public String en() {
            return String.format("%sCopied:%s %s => %s%s",
                    green(), reset(),
                    sourceKey.key(),
                    targetKey.key(),
                    eraseToEndOfScreen());
        }
        @Override
        public String enBatch() {
            return String.format("Copied: %s => %s",
                    sourceKey.key(), targetKey.key());
        }
    }
    static ConsoleOut.WithBatchMode deleteComplete(RemoteKey remoteKey) {
        return new DeleteComplete(remoteKey);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class DeleteComplete implements WithBatchMode {
        private final RemoteKey remoteKey;
        @Override
        public String en() {
            return String.format("Deleted: %s", remoteKey);
        }
        @Override
        public String enBatch() {
            return String.format("%sDeleted%s: %s%s",
                    green(), reset(),
                    remoteKey,
                    eraseToEndOfScreen());
        }
    }
    static ConsoleOut.WithBatchMode errorQueueEventOccurred(
            StorageEvent.ActionSummary actionSummary,
            Throwable error
    ) {
        return new ErrorQueueEventOccurred(actionSummary, error);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class ErrorQueueEventOccurred implements WithBatchMode {
        private final StorageEvent.ActionSummary actionSummary;
        private final Throwable error;
        @Override
        public String en() {
            return String.format("%s failed: %s: %s",
                    actionSummary.name(),
                    actionSummary.keys(),
                    error.getMessage());
        }
        @Override
        public String enBatch() {
            return String.format("%sERROR%s: %s %s: %s%s",
                    red(), reset(),
                    actionSummary.name(),
                    actionSummary.keys(),
                    error.getMessage(),
                    eraseToEndOfScreen());
        }
    }

}
